scanner/scanner.go

```diff
@@ -0,0 +1,150 @@
+package scanner
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"time"
+
+	"github.com/SkynetLabs/malware-scanner/clamav"
+	"github.com/SkynetLabs/malware-scanner/database"
+	"github.com/SkynetLabs/skynet-accounts/build"
+	"github.com/sirupsen/logrus"
+	"gitlab.com/NebulousLabs/errors"
+)
+
+var (
+	// sleepBetweenScans defines how long the scanner should sleep after
+	// scanning the DB and not finding any skylinks to scan.
+	sleepBetweenScans = build.Select(
+		build.Var{
+			Dev:      30 * time.Second,
+			Testing:  100 * time.Millisecond,
+			Standard: 10 * time.Second,
+		},
+	).(time.Duration)
+)
+
+// Scanner provides a convenient interface for working with ClamAV
+type Scanner struct {
+	staticCtx    context.Context
+	staticDB     *database.DB
+	staticClam   *clamav.ClamAV
+	staticLogger *logrus.Logger
+}
+
+// New returns a new Scanner with the given parameters.
+func New(ctx context.Context, db *database.DB, clam *clamav.ClamAV, logger *logrus.Logger) *Scanner {
+	return &Scanner{
+		staticCtx:    ctx,
+		staticDB:     db,
+		staticClam:   clam,
+		staticLogger: logger,
+	}
+}
+
+// SweepAndScan sweeps the DB for new skylinks, locks them, scans them,
+// and updates their records in the DB. The method returns a boolean flag that
+// instructs the caller to sleep after this call when there are no records to
+// be scanned.
+func (s Scanner) SweepAndScan(abort chan bool) (sleepAfter bool) {
+	sl, err := s.staticDB.SweepAndLock(s.staticCtx)
+	if err != nil && !errors.Contains(err, database.ErrNoDocumentsFound) {
+		s.staticLogger.Warnf("error while trying to lock a new record: %s", err)
+		return
+	}
+	if errors.Contains(err, database.ErrNoDocumentsFound) {
+		// Sleep after this run.
+		return true
+	}
+	if sl.Skylink == "" {
+		s.staticLogger.Warnf("empty skylink in record %+v", sl)
+		build.Critical(fmt.Sprintf("SweepAndLock returned a record with an empty skylink. Record hash: %s", hex.EncodeToString(sl.Hash[:])))
+		return
+	}
+	inf, desc, size, scannedSize, err := s.staticClam.ScanSkylink(sl.Skylink, abort)
+	if err != nil {
+		// Scanning failed, log the error and unlock the record for another attempt.
+		s.staticLogger.Debugln(errors.AddContext(err, "scanning failed"))
+		sl.Status = database.SkylinkStatusNew
+		sl.Timestamp = time.Now().UTC()
+		err = s.staticDB.SkylinkSave(s.staticCtx, sl)
+		if err != nil {
+			s.staticLogger.Debugln(errors.AddContext(err, "unlocking a skylink failed"))
+		}
+		return
+	}
+	// Sanity check: scannedSize vs size.
+	if scannedSize > size {
+		build.Critical(fmt.Sprintf("Scanned size (%d bytes) is more than the content size (%d bytes) for skylink %s", scannedSize, size, sl.Skylink))
```

Comment:
should we return or do some sort of handling of the size? if we are using a `build.Critical` we typically then handling the error and/or return to ensure the following code doesn't execute, or if it does it doesn't cause issues.

Reply:
Not really. I changed it to a warning. We want to know if it happens, because it's not expected but it won't break the execution.